include/swift/AST/DiagnosticEngine.h

```diff
@@ -724,6 +725,11 @@ namespace swift {
     /// diagnostic message.
     std::unique_ptr<diag::LocalizationProducer> localization;
 
+    /// This allocator will retain localized diagnostic strings containing the
+    /// diagnostic's message and identifier for the duration of compiler invocation.
+    llvm::BumpPtrAllocator localizationAllocator;
+    llvm::StringSaver localizationSaver;
```

Comment:
I don't think we need this here just to pass it down to localization. Localization, if enabled, is already references for the whole duration of `DiagnosticEgine`'s life so, we can just initialize it internally in `LocationProducer` and avoid exposing that detail to the caller.